File: src/mnemonic/converter/tlg/lzss.py
"""LZSS解凍アルゴリズムモジュール

TLG5形式で使用されるLZSS圧縮データの解凍機能を提供する。
"""

from typing import Protocol


class LZSSDecoderProtocol(Protocol):
    """LZSS解凍インターフェース

    LZSS圧縮されたバイト列を解凍するためのプロトコル定義。
    """

    def decode(self, data: bytes, output_size: int) -> bytes:
        """LZSS圧縮データを解凍する

        Args:
            data: LZSS圧縮されたバイト列
            output_size: 解凍後の期待サイズ（バイト）

        Returns:
            解凍されたバイト列

        Raises:
            ValueError: 不正な圧縮データの場合
        """
        ...


class LZSSDecoder:
    """LZSS解凍クラス

    TLG5形式で使用されるLZSS圧縮データを解凍する。
    スライディングウィンドウサイズは4096バイト。
    """

    WINDOW_SIZE: int = 4096
    """スライディングウィンドウのサイズ（バイト）"""

    MATCH_MAX_LENGTH: int = 18
    """最大マッチ長"""

    MATCH_MIN_LENGTH: int = 3
    """最小マッチ長"""

    def decode(self, data: bytes, output_size: int) -> bytes:
        """LZSS圧縮データを解凍する

        TLG5形式のLZSS解凍を行う。4096バイトのスライドバッファを使用。

        Args:
            data: LZSS圧縮されたバイト列
            output_size: 解凍後の期待サイズ（バイト）

        Returns:
            解凍されたバイト列

        Raises:
            ValueError: 不完全または不正な圧縮データの場合
        """
        if output_size == 0:
            return b""

        output = bytearray(output_size)
        # スライドバッファ（0で初期化）
        slide = bytearray(self.WINDOW_SIZE)
        slide_pos = 0

        output_pos = 0
        input_pos = 0
        data_len = len(data)

        while output_pos < output_size:
            if input_pos >= data_len:
                raise ValueError("不完全な圧縮データ: フラグバイトが不足しています")

            flags = data[input_pos]
            input_pos += 1

            for bit in range(8):
                if output_pos >= output_size:
                    break

                if flags & (1 << bit):
                    # TLG5: ビット1 = マッチ（バックリファレンス）
                    if input_pos + 2 > data_len:
                        raise ValueError("不完全な圧縮データ: マッチ情報が不足しています")

                    # 2バイトからマッチ位置と長さを抽出
                    low_byte = data[input_pos]
                    high_byte = data[input_pos + 1]
                    input_pos += 2

                    mpos = low_byte | ((high_byte & 0x0F) << 8)
                    mlen = ((high_byte >> 4) & 0x0F) + self.MATCH_MIN_LENGTH

                    # mlen == 18の場合は追加バイトを読む
                    if mlen == 18:
                        if input_pos >= data_len:
                            raise ValueError("不完全な圧縮データ: 追加長バイトが不足しています")
                        mlen += data[input_pos]
                        input_pos += 1

                    # スライドバッファからコピー
                    for _ in range(mlen):
                        if output_pos >= output_size:
                            break
                        byte_val = slide[mpos]
                        output[output_pos] = byte_val
                        slide[slide_pos] = byte_val
                        output_pos += 1
                        mpos = (mpos + 1) & (self.WINDOW_SIZE - 1)
                        slide_pos = (slide_pos + 1) & (self.WINDOW_SIZE - 1)
                else:
                    # TLG5: ビット0 = リテラルバイト
                    if input_pos >= data_len:
                        raise ValueError("不完全な圧縮データ: リテラルバイトが不足しています")
                    byte_val = data[input_pos]
                    input_pos += 1
                    output[output_pos] = byte_val
                    slide[slide_pos] = byte_val
                    output_pos += 1
                    slide_pos = (slide_pos + 1) & (self.WINDOW_SIZE - 1)

        return bytes(output)

File: internal/converter/tlg/lzss.go
// Package tlg は吉里吉里2 TLG画像フォーマット（TLG5/TLG6）のデコード機能を提供する。
package tlg

import (
	"errors"
	"fmt"
)

// LZSS解凍で使用する定数群。
//
// TLG5のブロック内圧縮データは吉里吉里方式のLZSSでエンコードされており、
// 4096バイトのスライド辞書・最大マッチ長18バイト・最小マッチ長3バイトという
// 固定パラメータを持つ（フォーマット仕様であり構成不可）。
const (
	WindowSize     = 4096
	MatchMaxLength = 18
	MatchMinLength = 3
)

// ErrLZSSIncompleteData はLZSS圧縮データが指定されたoutputSizeを満たす前に
// 尽きた場合のエラー（フラグバイト・マッチ情報・リテラルバイト・追加長バイトの
// いずれかが不足）。
var ErrLZSSIncompleteData = errors.New("不完全な圧縮データです")

// LZSSDecoder はTLG5で使用される吉里吉里方式LZSS圧縮データを解凍する。
// スライド辞書（4096バイト）とリングポインタ（slidePos）を状態として保持し、
// 同一インスタンスへの複数回のDecode呼び出しをまたいで辞書を持続させる。
//
// why not(辞書をDecodeローカルではなくフィールドで持つ理由): krkrz（吉里吉里
// ネイティブC++実装）のTVPLoadTLG5ローダーは、スライド辞書text[4096]とリング
// ポインタrを画像1枚のデコード（全ブロック×全チャンネル）を通じて1つだけ確保し、
// TVPTLG5DecompressSlideの呼び出しごとにrを引き回して辞書状態を持続させる。実機の
// RGBA .tlgアセットはブロック境界をまたぐバックリファレンス（前のチャンクで展開した
// バイトを後のチャンクが参照する）を含むため、辞書をチャンク（ブロック×チャンネル）
// ごとにゼロ初期化するとバックリファレンスが解決できずデコード結果がノイズになる。
// 移植元のPython参照実装はDecodeローカルで毎回スライド辞書を確保しており、この
// krkrz準拠の持続性を再現できていなかった（実アセットでノイズになることを実証済み）。
// 本Go実装はここで意図的にPython参照実装から乖離し、辞書状態をフィールドとして
// 保持する。1枚の画像に対しては呼び出し側（tlg5.go）が新しいLZSSDecoderを1つだけ
// 生成して全チャンクへ通すことで、画像間の辞書汚染を防ぐ。
//
// why not: LZSS解凍は外部I/Oを持たない純粋なアルゴリズムであり差し替え対象
// でもないため、lang-go.mdの「外部依存はインターフェース化」は適用対象外。
// 具象型として実装する（image.goのTLGImageDecoder等、既存コードベースの
// アルゴリズム実装パターンに合わせる）。
type LZSSDecoder struct {
	slide    [WindowSize]byte
	slidePos int
}

// NewLZSSDecoder はLZSSDecoderを初期化する。スライド辞書はゼロ初期化される
// （krkrz TVPLoadTLG5のmemset(text, 0, 4096)・r=0に相当）。
func NewLZSSDecoder() *LZSSDecoder {
	return &LZSSDecoder{}
}

// Decode はdataをLZSS解凍し、outputSizeバイトの解凍結果を返す。
//
// フラグバイト1バイトにつき8ビット分の制御を読み、ビット1=マッチ
// （2バイトのバックリファレンス。長さが18になる場合は追加の1バイトを読む）、
// ビット0=リテラル1バイトとして展開する。展開結果は同時にレシーバの4096バイトの
// スライド辞書へ書き戻され、以降のマッチ参照元になる。
//
// 辞書とリングポインタはレシーバに保持されるため、同一インスタンスへの連続した
// Decode呼び出しでは前回の呼び出しが残した辞書状態が引き継がれる（krkrzのrの
// スレッディングに相当）。マッチ位置mposは辞書全体（前チャンク由来のバイトを含む）を
// 参照できる。
func (d *LZSSDecoder) Decode(data []byte, outputSize int) ([]byte, error) {
	if outputSize == 0 {
		return []byte{}, nil
	}

	output := make([]byte, outputSize)
	slide := &d.slide
	slidePos := d.slidePos

	outputPos := 0
	inputPos := 0
	dataLen := len(data)

	for outputPos < outputSize {
		if inputPos >= dataLen {
			return nil, fmt.Errorf("%w: フラグバイトが不足しています", ErrLZSSIncompleteData)
		}

		flags := data[inputPos]
		inputPos++

		for bit := range 8 {
			if outputPos >= outputSize {
				break
			}

			if flags&(1<<uint(bit)) != 0 {
				// TLG5: ビット1 = マッチ（バックリファレンス）
				if inputPos+2 > dataLen {
					return nil, fmt.Errorf("%w: マッチ情報が不足しています", ErrLZSSIncompleteData)
				}

				lowByte := data[inputPos]
				highByte := data[inputPos+1]
				inputPos += 2

				mpos := int(lowByte) | (int(highByte&0x0F) << 8)
				mlen := int((highByte>>4)&0x0F) + MatchMinLength

				if mlen == MatchMaxLength {
					if inputPos >= dataLen {
						return nil, fmt.Errorf("%w: 追加長バイトが不足しています", ErrLZSSIncompleteData)
					}

					mlen += int(data[inputPos])
					inputPos++
				}

				for range mlen {
					if outputPos >= outputSize {
						break
					}

					b := slide[mpos]
					output[outputPos] = b
					slide[slidePos] = b
					outputPos++
					mpos = (mpos + 1) & (WindowSize - 1)
					slidePos = (slidePos + 1) & (WindowSize - 1)
				}
			} else {
				// TLG5: ビット0 = リテラルバイト
				if inputPos >= dataLen {
					return nil, fmt.Errorf("%w: リテラルバイトが不足しています", ErrLZSSIncompleteData)
				}

				b := data[inputPos]
				inputPos++
				output[outputPos] = b
				slide[slidePos] = b
				outputPos++
				slidePos = (slidePos + 1) & (WindowSize - 1)
			}
		}
	}

	// why not(エラー時にslidePosを書き戻さない): 解凍が途中でエラー終了した場合、
	// 呼び出し側（tlg5.go）は画像デコード全体を中断しこのLZSSDecoderを破棄する
	// ため、中途半端な辞書状態を書き戻す意味が無い。成功時のみ次チャンクへ引き継ぐ。
	d.slidePos = slidePos

	return output, nil
}
